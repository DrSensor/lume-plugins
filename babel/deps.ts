export * from "https://esm.sh/@babel/core@7";

export { default as presetEnv } from "https://esm.sh/@babel/preset-env@7";

export { default as pluginTransformRuntime } from "https://esm.sh/@babel/plugin-transform-runtime@7";
export { default as pluginProposalDecorators } from "https://esm.sh/@babel/plugin-proposal-decorators@7";
