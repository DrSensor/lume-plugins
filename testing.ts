import type { Page, Site, SourceMap } from "lume/core.ts";
import { assertSnapshot } from "std/testing/snapshot.ts";
import { basename } from "std/path/mod.ts";
import { printError } from "lume/core/errors.ts";

const cwUrl = import.meta.resolve("./");

/** Build a site and print errors */
export async function build(site: Site) {
  site.addEventListener("beforeSave", () => false); // Don't save the site to disk
  try {
    await site.build();
  } catch (error) {
    printError(error);
    throw error;
  }
}

function normalizeContent(
  content: string | Uint8Array | undefined,
): string | undefined {
  if (content instanceof Uint8Array) {
    return `Uint8Array(${content.length})`;
  }
  if (typeof content === "string") {
    return content // Normalize line ending for Windows
      .replaceAll("\r\n", "\n")
      .replaceAll(/base64,[^"]+/g, "base64,(...)");
  }
}

async function assertPageSnapshot(
  context: Deno.TestContext,
  page: Page,
) {
  let { data } = page;
  const { dest, src, content } = page;

  // Sort data alphabetically
  const entries = Object.entries(data).sort((a, b) => a[0].localeCompare(b[0]));
  data = Object.fromEntries(entries);

  await assertSnapshot(context, { src, dest, data, content });
}

export async function assertSiteSnapshot(
  context: Deno.TestContext,
  site: Site,
) {
  const { pages, files } = site;

  // Test number of pages
  await assertSnapshot(context, pages.length);

  // TODO: test site configuration
  await assertSnapshot(
    context,
    {
      formats: Array.from(site.formats.entries.values()).map((format) => {
        return {
          ...format,
          engines: format.engines?.length,
        };
      }),
    },
  );

  // Sort pages and files alphabetically
  pages.sort((a, b) => {
    return compare(a.src.path, b.src.path) || compare(a.dest.path, b.dest.path);
  });

  files.sort((a, b) => {
    return compare(a.src, b.src);
  });

  // Normalize some dynamic data
  pages.forEach((page) => {
    // Normalize data
    if (page.data.date instanceof Date) {
      page.data.date = new Date(0);
    }
    // Ignore comp object
    if (page.data.comp) {
      page.data.comp = {};
    }
    // Remove page reference
    page.data.page = undefined;

    // Normalize source maps
    if (page.data.sourceMap) {
      normalizeSourceMap(page.data.sourceMap as SourceMap);
    }

    // Remove pagination results details from the data
    if (Array.isArray(page.data.results)) {
      page.data.results = page.data.results.length;
    }
    // Remove alternates values (added by multilanguage plugin)
    if (page.data.alternates) {
      page.data.alternates = Object.keys(
        page.data.alternates as Record<string, Page>,
      );
    }
    // Remote base path because it's different in the test environment
    page.src.remote = page.src.remote?.replace(cwUrl, "");
    delete page.src.created;
    delete page.src.lastModified;

    // Normalize content for Windows
    page.content = normalizeContent(page.content);
    page.data.content = normalizeContent(
      page.data.content as string | Uint8Array | undefined,
    );

    // Source maps
    if (page.dest.ext === ".map") {
      const map = JSON.parse(page.content as string);
      normalizeSourceMap(map);
      page.content = JSON.stringify(map);
      page.data.content = JSON.stringify(map);
    }
  });

  // Test static files
  await assertSnapshot(
    context,
    files.map((file) => {
      // Remote base path because it's different in the test environment
      file.remote = file.remote?.replace(cwUrl, "");
      return file;
    }),
  );

  // Test pages
  for (const page of pages) {
    await assertPageSnapshot(context, page);
  }
}

function compare(a: string, b: string): number {
  return a > b ? 1 : a < b ? -1 : 0;
}

function normalizeSourceMap(sourceMap: SourceMap) {
  sourceMap.sourceRoot = sourceMap.sourceRoot
    ? basename(sourceMap.sourceRoot)
    : undefined;
  sourceMap.file = sourceMap.file ? basename(sourceMap.file) : undefined;
  sourceMap.sources = sourceMap.sources.map((source: string) =>
    basename(source)
  );
}
