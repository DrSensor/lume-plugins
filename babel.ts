import { merge } from "lume/core/utils.ts";
import { toFileUrl } from "lume/deps/path.ts";
import { prepareAsset } from "lume/plugins/source_maps.ts";

import type { Site } from "lume/core.ts";
import type { Options } from "./babel/options.ts";

import * as babel from "./babel/deps.ts";
import defaultOptions from "./babel/options.ts";

export default function (userOptions?: Partial<Options>) {
  let { extensions, ...babelOptions } = merge(defaultOptions, userOptions);
  if (userOptions?.plugins) {
    babelOptions.plugins!.unshift(...defaultOptions.plugins!);
  }
  babelOptions = babel.loadOptions(babelOptions)!;
  return (site: Site) => {
    site.loadAssets(extensions);
    site.process(extensions, async (page) => {
      const script = prepareAsset(site, page);

      // WARNING: Passing cached plugin instances is not supported in babel.loadPartialConfig()
      // const config = await babel.loadPartialConfigAsync(babelOptions!);
      // babelOptions = config!.options;

      if (script.enableSourceMap) {
        babelOptions.sourceMaps ??= true;
        babelOptions.sourceFileName = toFileUrl(script.filename).href;
      }

      const result = await babel.transformAsync(script.content, babelOptions)!;
      page.content = result!.code!;
      page.updateDest({ ext: ".js" });
    });
  };
}
