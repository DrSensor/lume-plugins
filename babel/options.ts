import * as babel from "./deps.ts";

export default {
  extensions: babel.DEFAULT_EXTENSIONS,
  sourceType: "unambiguous",
  presets: [babel.presetEnv],
  plugins: [
    babel.pluginTransformRuntime,
    [babel.pluginProposalDecorators, { version: "2022-03" }],
  ],
  targets: {
    esmodules: true,
    browsers: "unreleased Chrome versions",
  },
} as Options;

export interface Options extends babel.TransformOptions {
  /** Filter which file extensions should be transpiled by Babel.
  @see https://babeljs.io/docs/en/babel-core#default_extensions */
  extensions: string[];

  /** @see https://babeljs.io/docs/en/options#output-targets */
  targets?: // adding ðŸ‘ˆ cuz somehow babel.InputOptions is missing ðŸ¤”
    | string
    | string[]
    | Partial<
      {
        esmodules: boolean;
        browsers: string | string[];
        node: string | "current" | true;
        safari: string | "tp";
      }
    >
    | Record<
      | "android"
      | "chrome"
      | "edge"
      | "electron"
      | "firefox"
      | "ie"
      | "ios"
      | "node"
      | "opera"
      | "rhino"
      | "safari"
      | "samsung",
      `${number}`
    >;
}
