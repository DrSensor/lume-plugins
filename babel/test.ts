import type { SiteOptions } from "lume/core.ts";
import lume from "lume/mod.ts";
import * as path from "std/path/mod.ts";
import * as test from "../testing.ts";
import babel from "../babel.ts";

const cwd = path.fromFileUrl(import.meta.resolve("./"));
const options: Partial<SiteOptions> = {
  cwd: path.join(cwd, "demo"),
  quiet: true,
};

Deno.test("babel plugin", async (t) => {
  const site = lume(options);

  site.use(babel());

  await test.build(site);
  await test.assertSiteSnapshot(t, site);
});
