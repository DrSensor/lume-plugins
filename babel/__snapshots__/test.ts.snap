export const snapshot = {};

snapshot[`babel plugin 1`] = `1`;

snapshot[`babel plugin 2`] = `
{
  formats: [
    {
      engines: 1,
      ext: ".tmpl.ts",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".tmpl.js",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: undefined,
      ext: ".tmpl.json",
      pageLoader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".json",
    },
    {
      engines: 1,
      ext: ".md",
      pageLoader: [AsyncFunction: text],
    },
    {
      asset: true,
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      pageLoader: [AsyncFunction: text],
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
    },
    {
      componentLoader: [AsyncFunction: text],
      engines: 1,
      ext: ".njk",
      includesPath: "_includes",
      pageLoader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yaml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      asset: true,
      engines: undefined,
      ext: ".jsx",
      pageLoader: [AsyncFunction: text],
    },
    {
      asset: true,
      engines: undefined,
      ext: ".es6",
      pageLoader: [AsyncFunction: text],
    },
    {
      asset: true,
      engines: undefined,
      ext: ".es",
      pageLoader: [AsyncFunction: text],
    },
    {
      asset: true,
      engines: undefined,
      ext: ".mjs",
      pageLoader: [AsyncFunction: text],
    },
    {
      asset: true,
      engines: undefined,
      ext: ".cjs",
      pageLoader: [AsyncFunction: text],
    },
  ],
}
`;

snapshot[`babel plugin 3`] = `
[
]
`;

snapshot[`babel plugin 4`] = `
{
  content: 'class Title {
  #A = getComputedStyle(document.body).fontSize;

  get size() {
    return this.#A;
  }

  set size(v) {
    this.#A = v;
  }

  #text = "";

  get text() {
    return this.#text;
  }

  constructor(text) {
    this.#text = text.data;
  }

}

const titles = [];

for (const h1 of document.getElementsByTagName("h1")) {
  const title = new Title(h1.firstChild);
  title.size = getComputedStyle(h1).fontSize;
  titles.push(title);
}

console.debug(titles);',
  data: {
    content: 'class Title {
  accessor size = getComputedStyle(document.body).fontSize
  #text = ""
  get text() { return this.#text }
  constructor(text) { this.#text = text.data }
}

const titles = []
for (const h1 of document.getElementsByTagName("h1")) {
  const title = new Title(h1.firstChild)
  title.size = getComputedStyle(h1).fontSize
  titles.push(title)
}
console.debug(titles)
',
    date: 1970-01-01T00:00:00.000Z,
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [
    ],
    url: "/script.js",
  },
  dest: {
    ext: ".js",
    path: "/script",
  },
  src: {
    ext: ".js",
    path: "/script",
    remote: undefined,
  },
}
`;
